torch/csrc/autograd/variable.h

```diff
@@ -453,13 +453,22 @@ inline Variable make_variable_differentiable_view(
     CreationMeta creation_meta,
     c10::optional<std::function<at::Tensor(const at::Tensor&)>> view_func = c10::nullopt) {
   if (data.defined()) {
-    auto data_impl_copy = data.getIntrusivePtr()->shallow_copy_and_detach(
-      /*version_counter=*/0,
-      /*allow_tensor_metadata_change=*/true);
-    data_impl_copy->set_autograd_meta(std::make_unique<DifferentiableViewMeta>(
-      data_impl_copy.get(), std::move(base), std::move(view_func),
-      creation_meta));
-    return Variable(data_impl_copy);
+    c10::intrusive_ptr<at::TensorImpl> data_impl;
+    // If we already did a TensorImpl allocation for data, just reuse it.
+    // Otherwise(e.g tensor.swapdim(0, 0) when we return the same tensor as input),
+    // we have to use shallow_copy_and_detach to create a new TensorImpl to avoid
+    // moving leaf node into graph interior. This guarantees only 1 TensorImpl
+    // allocation happens in view ops.
+    if (base.unsafeGetTensorImpl() == data.unsafeGetTensorImpl()) {
+      data_impl = data.getIntrusivePtr()->shallow_copy_and_detach(
+        /*version_counter=*/0,
+        /*allow_tensor_metadata_change=*/true);
+    } else {
+       data_impl = data.getIntrusivePtr();
```

Comment:
Wondering if it is also true that `data.use_count() == 1` here

Reply:
Great point! That made me realize we were passing Tensor by value which increased the use_count unnecessarily in `as_view`. Updated the condition to rely on use_count. PTAL!